src/xdp/ebpfextension.c

```diff
@@ -0,0 +1,277 @@
+
+    if (Client == NULL) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    if (Provider->DetachCallback != NULL) {
+        Status = Provider->DetachCallback(Client);
+        ASSERT(NT_SUCCESS(Status));
+    }
+
+Exit:
+
+    TraceExitStatus(TRACE_CORE);
+    return Status;
+}
+
+static
+VOID
+EbpfExtensionProviderCleanup(
+    _Frees_ptr_ VOID *ProviderBindingContext
+    )
+{
+    TraceEnter(TRACE_CORE, "ProviderBindingContext=%p", ProviderBindingContext);
+
+    ExFreePoolWithTag(ProviderBindingContext, XDP_POOLTAG_EBPF_NMR);
+
+    TraceExitSuccess(TRACE_CORE);
+}
+
+VOID
+EbpfExtensionProviderUnregister(
+    _Frees_ptr_opt_ EBPF_EXTENSION_PROVIDER *ProviderContext
+    )
+{
+    TraceEnter(TRACE_CORE, "ProviderContext=%p", ProviderContext);
+
+    if (ProviderContext != NULL) {
+        NTSTATUS Status = NmrDeregisterProvider(ProviderContext->NmrProviderHandle);
+
+        if (Status == STATUS_PENDING) {
+            NmrWaitForProviderDeregisterComplete(ProviderContext->NmrProviderHandle);
+        }
+
+        ExFreePoolWithTag(ProviderContext, XDP_POOLTAG_EBPF_NMR);
+    }
+
+    TraceExitSuccess(TRACE_CORE);
+}
+
+NTSTATUS
+EbpfExtensionProviderRegister(
+    _In_ const NPIID *NpiId,
+    _In_ const EBPF_EXTENSION_PROVIDER_PARAMETERS *Parameters,
+    _In_opt_ EBPF_EXTENSION_ON_CLIENT_ATTACH *AttachCallback,
+    _In_opt_ EBPF_EXTENSION_ON_CLIENT_DETACH *DetachCallback,
+    _In_opt_ const VOID *CustomData,
+    _Outptr_ EBPF_EXTENSION_PROVIDER **ProviderContext
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    EBPF_EXTENSION_PROVIDER *Provider = NULL;
+    NPI_PROVIDER_CHARACTERISTICS *Characteristics;
+
+    TraceEnter(TRACE_CORE, "Parameters=%p", Parameters);
+
+    Provider = ExAllocatePoolZero(NonPagedPoolNx, sizeof(*Provider), XDP_POOLTAG_EBPF_NMR);
+    if (Provider == NULL) {
+        Status = STATUS_NO_MEMORY;
+        goto Exit;
+    }
+
+    Characteristics = &Provider->Characteristics;
+    Characteristics->Length = sizeof(NPI_PROVIDER_CHARACTERISTICS);
+    Characteristics->ProviderAttachClient = EbpfExtensionProviderAttachClient;
+    Characteristics->ProviderDetachClient = EbpfExtensionProviderDetachClient;
+    Characteristics->ProviderCleanupBindingContext = EbpfExtensionProviderCleanup;
+    Characteristics->ProviderRegistrationInstance.Size = sizeof(NPI_REGISTRATION_INSTANCE);
+    Characteristics->ProviderRegistrationInstance.NpiId = NpiId;
+    Characteristics->ProviderRegistrationInstance.NpiSpecificCharacteristics =
+        Parameters->ProviderData;
+    Characteristics->ProviderRegistrationInstance.ModuleId = Parameters->ProviderModuleId;
+
+    Provider->AttachCallback = AttachCallback;
+    Provider->DetachCallback = DetachCallback;
+    Provider->CustomData = CustomData;
+
+    Status = NmrRegisterProvider(Characteristics, Provider, &Provider->NmrProviderHandle);
+    if (!NT_SUCCESS(Status))
+        goto Exit;
```

Comment:
I think we're generally always wrapping in braces, even for one line? But more importantly, I think it's worth a log for any function XDP calls and fails.

Reply:
Yup, missed this during copy/paste, and added a log.